viewing_party/party.py

```diff
@@ -1,23 +1,181 @@
+import copy
+
 # ------------- WAVE 1 --------------------
 
 def create_movie(title, genre, rating):
-    pass
+    if not (title and genre and rating):
+        return None
+
+    movie = {
+        "title" : title, 
+        "genre" : genre, 
+        "rating" : rating
+    }
+
+    return movie
+
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+
+    return user_data
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+
+    return user_data
+
+def watch_movie(user_data, title):
+    new_user_data = {
+        "watchlist" : user_data["watchlist"].copy(),
+        "watched" : user_data["watched"].copy()
+    }
+    for movie in user_data["watchlist"]:
+        if title in movie["title"]:
+            new_user_data["watchlist"].remove(movie)
+            new_user_data["watched"].append(movie)
+
+    return new_user_data
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    sum_rating = 0
+    watched_movies = user_data["watched"]
+
+    if len(watched_movies) == 0: return 0.0
+
+    for movie in watched_movies:
+        sum_rating += movie["rating"]
+    average_rating = sum_rating / len(watched_movies)
+
+    return average_rating
+
+def get_most_watched_genre(user_data):
+    genre_dict = {}
+    highest_num = 0
+    highest_genre = ""
+    watched_movies = user_data["watched"]
+
+    if len(watched_movies) == 0: return None
+
+    for movie in watched_movies:
+        if movie["genre"] in genre_dict:
+            genre_dict[movie["genre"]] += 1
+        else:
+            genre_dict[movie["genre"]] = 1
+
+    for genre, num in genre_dict.items():
+        if num > highest_num:
+            highest_num = num
+            highest_genre = genre
+    
+    return highest_genre
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_friend_movies(friend_list):
+    friend_watched_movies = []
+
+    for movies in friend_list:
+        for movie in movies["watched"]:
+            friend_watched_movies.append(movie)
+    
+    return friend_watched_movies
+
+def get_unique_watched(user_data):
+    watched_movies = user_data["watched"].copy()
+    friend_watched_movies = get_friend_movies(user_data["friends"])
+    not_watched_by_friends = []
+
+    fw_copy = friend_watched_movies.copy()

```

Comment:
I see you made a copy of a copy here. Remember, making copies is really only needed _if_ we are iterating over a list while we need to mutate it by removing items in a standard for loop. Otherwise, no need to make extra variables that need to be stored in memory.

Are you using this `fw_copy` anywhere?

```suggestion

```